### ISSUE TITLE
Memory domination not working

### ISSUE THREAD

**Issue by domainexpert** [tracer-x/klee#265]
Regarding the algorithm implemented in `Dependency::getConcreteStore` to compute the memory dominated by another memory so it doesn't get considered in subsumption checks, `basic/domination.c` has no subsumptions. It should have if the algorithm actually works.

An explanation of memory domination is in #199, and an implementation was previously attempted in #231.

**EDIT: Added references to #199 and #231.**

**Comment by domainexpert** [comment]
@rasoolmaghareh We attempt to resolve this in #268.

**Comment by domainexpert** [comment]
@rasoolmaghareh This is not resolved yet in #268, so reopening.

**Comment by domainexpert** [comment]
The following example has been described in #203 as an example of memory allocation that need not be considered preventing subsumption. At the time, this was considered a symptom of the need to clear local memory when exiting the function. It is better to handle this as an example where memory domination is needed. Note that the header comment is outdated.
```
/*
 * An example that exhibits issue of not clearing local memory after
 * function exit in Tracer-X KLEE revision 08b728a, which results in
 * subsumption failure as defined memory in the interpolant is not
 * found in the memory of the state to be subsumed. This is because
 * some allocations in the state to be subsumed have not yet been
 * initialized, in particular, the allocation of the return value of
 * matchhere. The store to this allocation only happens shortly before
 * the return of the function, and therefore at most points during the
 * execution of the function, the value is undefined.
 *
 * In a stored interpolant, however, the do loop in matchhere may have
 * been executed for several iterations. This results in the return
 * value becomes defined (as it has been stored into before in some
 * iterations). It is included in the interpolant as the return value
 * is used in the if condition in the body of the do loop to prove
 * that control can only go one direction. This defined memory cannot
 * subsume the undefined memory in the state to be subsumed.
 *
 * The issue disappear when the "do" and "while" statements in
 * matchhere are removed.
 *
 * Simplified from simple regular expression matching in
 *   The Practice of Programming
 *   Brian W. Kernighan, Rob Pike
 *
 * The code with KLEE harness is obtained from the following
 * page of the KLEE tutorial on December 2015:
 *   http://klee.github.io/resources/Regexp.c.html
 */
#ifdef LLBMC
#include <llbmc.h>
#else
#include <klee/klee.h>
#endif

static int matchhere(char *re, char *text) {
  if (re[0] == '\0')
     return 0;
  if (re[1] == '*') {
    char c = re[0];
    re += 2;
    do {
      if (matchhere(re, text))
        return 1;
    } while (*text != '\0' && *text++ == c);
    return 0;
  }
  if (*text!='\0' && re[0]==*text)
    return matchhere(re+1, text+1);
  return 0;
}

/*
 * Harness for testing with KLEE.
 */

// The size of the buffer to test with.
#define SIZE 7

int main() {
  // The input regular expression.
  char re[SIZE];

// Make the input symbolic.
#ifdef LLBMC
  for (int i = 0; i < SIZE; ++i) {
    re[i] = __llbmc_nondef_char();
  }
#else
  klee_make_symbolic(re, sizeof re, "re");
#endif

  // Try to match against a constant string "hello".
  if (matchhere(re, "hello"))
    return 1;
  return 0;
}
```
